src/vs/workbench/services/workspaces/common/workspaceTrust.ts

```diff
@@ -340,6 +335,121 @@ export class WorkspaceTrustManagementService extends Disposable implements IWork
 	}
 }
 
+export class RemoteWorkspaceTrustManagementService extends AbstractWorkspaceTrustManagementService {
+
+	private _remoteAuthority!: ResolverResult;
+
+	constructor(
+		@IConfigurationService configurationService: IConfigurationService,
+		@IStorageService storageService: IStorageService,
+		@IUriIdentityService uriIdentityService: IUriIdentityService,
+		@IWorkbenchEnvironmentService environmentService: IWorkbenchEnvironmentService,
+		@IWorkspaceContextService workspaceService: IWorkspaceContextService,
+		@IRemoteAuthorityResolverService private readonly remoteAuthorityResolverService: IRemoteAuthorityResolverService,
+	) {
+		super(configurationService, storageService, uriIdentityService, environmentService, workspaceService, () => this.calculateWorkspaceTrust());
+
+		this._trustState.isTrusted = false;
+
+		// Remote - resolve remote authority
+		if (this.environmentService.remoteAuthority) {
+			this.remoteAuthorityResolverService.resolveAuthority(this.environmentService.remoteAuthority)
+				.then(async result => {
+					this._remoteAuthority = result;
+					await this.updateWorkspaceTrust();
+
+					// Show workspace trust modal dialog
+					this._onDidInitiateWorkspaceTrustRequestOnStartup.fire();
+				});
+		}
+	}
+
+	override canSetWorkspaceTrust(): boolean {
+		// Remote - remote authority not yet resolved, or remote authority explicitly sets workspace trust
+		if (this.environmentService.remoteAuthority && (!this._remoteAuthority || this._remoteAuthority.options?.isTrusted !== undefined)) {
+			return false;
+		}
+
+		return super.canSetWorkspaceTrust();
+	}
+
+	protected async calculateWorkspaceTrust(): Promise<boolean> {
+		const isTrustedDefault = this.calculateWorkspaceTrustDefault();
+		if (isTrustedDefault !== undefined) {
+			return isTrustedDefault;
+		}
+
+		// Remote - remote authority explicitly sets workspace trust
+		if (this.environmentService.remoteAuthority && this._remoteAuthority.options?.isTrusted !== undefined) {
+			return this._remoteAuthority.options.isTrusted;
+		}
+
+		// Workspace
+		const workspaceUris = await this.getWorkspaceUris();
+		const trusted = this.getUrisTrust(workspaceUris);
+
+		return trusted;
+	}
+
+	protected async getWorkspaceUris(): Promise<URI[]> {
+		const workspaceUris: URI[] = [];
```

Comment:
i see we only call `getCanonicalUris()` here and not in `getUriTrustInfo` or `getUrisTrust` but all of these are exposed. since any consumer could have a uri they want to know about the trust, maybe we should add or move the translation to that layer.

Reply:
Doh! I have refactored the code, and moved the translation to the lowest later.